### PR TITLE
Updates to our IOperation testing

### PR DESCRIPTION
**Pull request by jaredpar**
The IOperation verification includes running an extra validation walker over our trees for virtually every compilation done in our unit tests. This validation is not suitable for every test, due to the computations needed, and hence is disabled where necessary. 

Previously this ran via a conditional compilation approach. This made testing IOperation a bit awkward. Had to first do a recompile and then run a test. This PR changes the IOperation leg to instead look for an environment variable: `%ROSLYN_TEST_IOPERATION%`. 

I was a bit leery about using an environment variable here. But there really isn't any other supported mechanism for feeding options to tests via xunit. Other tests in our system have decided to use environment variables as well hence this isn't adding a new mechanism, rather unifying our tests. 

This won't yet move the IOperation testing into our normal PR flow. Want to get our Spanish tests stable first before we add another leg into the mix. 

Note: these changes have all already gone through full review. This is just merging features/optest into master.

### CODE REVIEW

**Comment by jaredpar** [comment]
FYI @dotnet/roslyn-compiler

File: src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs [hunk 1]
```diff
@@ -873,7 +873,7 @@ internal sealed partial class LazyBinaryOperatorExpression : BaseBinaryOperatorE
     internal abstract class BaseTupleBinaryOperatorExpression : Operation, ITupleBinaryOperation
     {
         public BaseTupleBinaryOperatorExpression(BinaryOperatorKind operatorKind, SemanticModel semanticModel, SyntaxNode syntax, ITypeSymbol type, Optional<object> constantValue, bool isImplicit)
-            : base(OperationKind.BinaryOperator, semanticModel, syntax, type, constantValue, isImplicit)
+            : base(OperationKind.TupleBinaryOperator, semanticModel, syntax, type, constantValue, isImplicit)
```
Comment (333fred) [inline src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs]: Seems like an unrelated bug fix, can we break it out into a separate change?

File: src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs [hunk 2]
```diff
@@ -310,6 +310,10 @@ private BoundForEachStatement BindForEachPartsWorker(DiagnosticBag diagnostics,
                         {
                             // Bind the expression for error recovery, but discard all new diagnostics
                             iterationErrorExpression = BindExpression(node.Variable, new DiagnosticBag());
+                            if (iterationErrorExpression.Kind == BoundKind.DiscardExpression)
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs]: Seems unrelated?

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs [hunk 3]
```diff
@@ -6749,6 +6749,13 @@ static System.Collections.Generic.IEnumerable<int> M()
                 //         foreach (_ in M())
                 Diagnostic(ErrorCode.ERR_MustDeclareForeachIteration, "_").WithLocation(6, 18)
                 );
+
+            var tree = comp.SyntaxTrees.Single();
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs]: Seems unrelated?

File: src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs (hunk 1 above)
Comment (333fred) [reply src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs]: Ah, nevermind, I see that these were in a separate commit. To satisfy my curiosity, why were these bug fixes made in a separate branch, and not in master directly?

---
In reply to: [188370704](https://github.com/dotnet/roslyn/pull/26862#discussion_r188370704) [](ancestors = 188370704)

Comment (jaredpar) [reply src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs]: I actually thought they'd take longer and be more disruptive. Hence we did them in a branch instead. Turns out it was much easier than I thought based on the number of failures we started with.
